syntax = "proto3";

package boxes;

// ---------------------------------------------------------------------------------------------------------------------

service DataSource {
  // List available tables
  rpc GetTableNames ()            returns (TableNames) {}

  // Get table schema
  rpc GetTableSchema (TableName)  returns (TableSchema) {}

  // Get table contents
  rpc GetTable (TableName)        returns (Table) {}
}
  
message TableName {
  string n = 1;
}

message TableNames {
  repeated string n = 1;
}

message TableSchema {
  repeated ColumnDefinition column_definitions = 1;
}

enum ColumnType {
  UNUSED  = 0;
  DECIMAL = 1;
  STRING  = 2;
}

message ColumnDefinition {
  string      name = 1;
  ColumnType  type = 2;
}

message Table {
  repeated ColumnValues columns = 1;
}

message ColumnValues {
  repeated string str_values = 1;
  repeated double dec_values = 2;
}

// ---------------------------------------------------------------------------------------------------------------------

service Algorithms {
  rpc Max (DoubleList)      returns (DoubleValue) {}
  rpc Min (DoubleList)      returns (DoubleValue) {}
  rpc Average (DoubleList)  returns (DoubleValue) {}
  rpc Median (DoubleList)   returns (DoubleValue) {}
}

message DoubleList {
  repeated double values = 1;
}

message DoubleValue {
  double value = 1;
}

// ---------------------------------------------------------------------------------------------------------------------

// 
service DataEngine {
  // Data engine känner till alla tabeller i alla connectorer
  // Den håller en tabell med alla tabeller och huruvida de är laddade eller ej
  // Den laddar tabeller till en sqlite in memory databas vid behov, mha namnkonvention borde vi kunna hitta tabeller i queries

  // Plot x antal axlar
  // Axel kategorisk eller numerisk
  rpc ExecuteQuery (Query) returns (Table) {}
}


message Query {
  string q = 1;
}

